File: primalitytest.py
import sys
sys.stdin = open("in.txt","rb")

from random import randint
def exp(a,b,m):
    ans=1
    a=a%m
    while(b>0):
        if(b&1):
            ans=(ans*a)%m
        b=b>>1##b/=2
        a=(a*a)%m
    return ans

def isPrimeUsingFermat(p,iter):
    if p==1:
        return False
    else:
        for i in range(0,iter):
            a = randint(0,p-2)+1
            if(exp(a,p-1,p)!=1):
                return False
        return True
        
t = int(raw_input())
while(t):
    x=int(raw_input())
    if(isPrimeUsingFermat(x,20)):
        print "YES"
    else:
        print "NO"
    t=t-1

File: algos_implementation/primalitytest.cpp
#include <cstdio>
#include <cstdlib>
#define toi long long int

/*evaluate (a^b)mod n*/
/*Culprit!! stops overflow only if a*a doesnt exceed long long int*//
toi exp(toi a,toi b,toi mod)
{
    toi ans = 1;
    a = a % mod;
    while(b>0)
    {
        if(b&1)
            ans = (ans * a) % mod;
        b=b>>1;//b/2
        a=(a*a) % mod;
    }    
    return ans;
}
/* Fermat's Little Theorem For prime p and a coprime to p. a^(p-1) % p = 1 
   Fails for Carmichael number as as they are odd composite numbers who satisfy this property e.g
*/
bool isPrimeUsingFermat(toi p,toi iter)
{
    if(p==1)
        return false;
    for(toi i=0;i<iter;i++)
    {
        toi a = rand()%(p-1)+1; // ensuring a and p are coprime
        if(exp(a,p-1,p)!=1)
            return false;
    }
    return true;
}
int main()
{
    int i,x;
    freopen("in.txt","r",stdin);
    for(scanf("%d",&i);i--;)
    {
        scanf("%d",&x);
        if(isPrimeUsingFermat(x,5000))
            printf("Yes\n");
        else
            printf("No\n");
    }
}